File: services/task_service.py
from services.db_service import get_db
from services._query_utils import _build_conditions


def _build_task_conditions(status, priority, owner_role):
    return _build_conditions({
        "status": status,
        "priority": priority,
        "owner_role": owner_role,
    })


def get_tasks(conn=None, status=None, priority=None, owner_role=None, limit=100):
    where, params = _build_task_conditions(status, priority, owner_role)
    params.append(limit)

    should_close = conn is None
    if conn is None:
        conn = get_db()
    try:
        cur = conn.execute(f"""
            SELECT task_id, task_title, task_description, status, priority,
                   owner_role, owner_user_id, due_at, created_at, completed_at,
                   feedback, recommendation_id, event_id, target_object_type,
                   target_object_id
            FROM action_tasks {where} ORDER BY created_at DESC LIMIT ?
        """, params)
        return [dict(r) for r in cur.fetchall()]
    finally:
        if should_close:
            conn.close()


def get_tasks_with_count(conn=None, status=None, priority=None, owner_role=None, limit=100):
    should_close = conn is None
    if conn is None:
        conn = get_db()
    try:
        items = get_tasks(conn, status, priority, owner_role, limit)
        where, count_params = _build_task_conditions(status, priority, owner_role)
        total = conn.execute(
            f"SELECT COUNT(*) FROM action_tasks {where}", count_params
        ).fetchone()[0]
        return items, total
    finally:
        if should_close:
            conn.close()


def get_task_by_id(conn=None, task_id=None):
    should_close = conn is None
    if conn is None:
        conn = get_db()
    try:
        cur = conn.execute("""
            SELECT task_id, recommendation_id, event_id, task_title,
                   task_description, task_source, owner_role, owner_user_id,
                   priority, due_at, status, feedback, completed_at, created_at
            FROM action_tasks WHERE task_id = ?
        """, (task_id,))
        row = cur.fetchone()
        return dict(row) if row else None
    finally:
        if should_close:
            conn.close()

File: internal/service/task_service.go
package service

import (
	"context"
	"fmt"

	"baxi/internal/model"
	taskRepo "baxi/internal/repository/task"
)

// TaskService handles business logic for task-related operations.
type TaskService struct {
	repo *taskRepo.Repository
}

// NewTaskService creates a new TaskService.
func NewTaskService(repo *taskRepo.Repository) *TaskService {
	return &TaskService{repo: repo}
}

// ListTasks retrieves tasks with optional filters and pagination,
// mapping repository rows to model types.
func (s *TaskService) ListTasks(
	ctx context.Context,
	filters model.TaskFilters,
	limit, offset int,
) (*model.TaskListResponse, error) {
	repoFilters := taskRepo.TaskFilters{
		Status:   filters.Status,
		Priority: filters.Priority,
		Owner:    filters.Owner,
	}

	rows, total, err := s.repo.ListTasks(ctx, repoFilters, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("list tasks: %w", err)
	}

	items := make([]model.Task, 0, len(rows))
	for _, row := range rows {
		items = append(items, mapRowToTask(row))
	}

	return &model.TaskListResponse{
		Items: items,
		Total: total,
	}, nil
}

// mapRowToTask converts a repository row to a model Task.
// alert_id in PostgreSQL maps to event_id in the model.
func mapRowToTask(row taskRepo.TaskRow) model.Task {
	desc := ""
	if row.TaskDescription != nil {
		desc = *row.TaskDescription
	}

	ownerRole := ""
	if row.OwnerRole != nil {
		ownerRole = *row.OwnerRole
	}

	priority := row.Priority
	if priority == "" {
		priority = "medium"
	}

	status := row.Status
	if status == "" {
		status = "todo"
	}

	var eventID *string
	if row.AlertID != nil {
		e := *row.AlertID
		eventID = &e
	}

	return model.Task{
		TaskID:           row.TaskID,
		TaskTitle:        row.TaskTitle,
		TaskDescription:  desc,
		Status:           status,
		Priority:         priority,
		OwnerRole:        ownerRole,
		OwnerUserID:      row.OwnerUserID,
		DueAt:            row.DueAt,
		CreatedAt:        row.CreatedAt,
		CompletedAt:      row.CompletedAt,
		Feedback:         row.Feedback,
		RecommendationID: row.RecommendationID,
		EventID:          eventID,
		TargetObjectType: row.TargetObjectType,
		TargetObjectID:   row.TargetObjectID,
	}
}
